### PR TITLE
Update to python3

### PR DESCRIPTION
**Pull request by Cypher1**
As Python2 is no longer maintained, this change is an attempt to move naive to use Python3.

Note: Tests currently fail on my machine, but I assume this is due to my use of WSL2 or a misconfiguration

An example failure:
```
test 'tests/ncc/void' failed:
compilation failed with stderr:
    Relocations for that section are not supported (found rela section .rela.data.rel.local)
    Relocations for that section are not supported (found rela section .rela.data.rel.local)
    Relocations for that section are not supported (found rela section .rela.data.rel.local)
    Relocations for that section are not supported (found rela section .rela.rodata)
    Unsupported relocation type: 4
    ncc: src/elf.c:1194: link_elf_executable: Assertion `false' failed.
```

### CODE REVIEW

**Comment by orodley** [comment]
Oh thanks, I'd been vaguely intending to do this at some point. I'll try to take a look some time this week.

That failure looks like a known issue, see https://github.com/orodley/naive/issues/2#issuecomment-388658140

**Comment by orodley** [comment]
Can you rebase on master and try running the tests again? 825efc6 should hopefully have fixed the issue.

**Comment by Cypher1** [comment]
Thanks, that did it! I've fixed up the test code to handle the bytes vs strings properly. Look forward to your review

File: meta/peg.py [hunk 1]
```diff
@@ -68,15 +68,15 @@ def at(self, offset):
...
+        #print(self.at(0))
         self.position += 1
         return self.at(-1)
 
     def read_toplevel(self):
         name = self.read_token()
         equals = self.read_token()
 
-        assert all(ident_char(c) for c in name)
+        assert all([ident_char(c) for c in name])
```
Comment (orodley) [inline meta/peg.py:79]: Adding the square brackets here shouldn't be necessary - Python 2 and 3 both support generator expressions.

Same thing applies in other places where you've added them.

File: run_tests.py [hunk 2]
```diff
@@ -93,16 +93,17 @@ def start_compiling(test_dir):
...
+    testcase.expected_compile_stderr = b''
+    testcase.expected_run_stdout = b''
+    testcase.expected_run_stderr = b''
+    testcase.run_stdin = b''
 
     test_filenames = []
     extra_flags = []
     for filename in sub_files:
-        with open(os.path.join(test_dir, filename), 'r') as f:
+        print(test_dir, filename)
```
Comment (orodley) [inline run_tests.py:105]: Looks like a debug print left in by accident?

**Review by orodley — CHANGES_REQUESTED** [review]
Looks good other than a few minor comments.

File: meta/peg.py (hunk 1 above)
Comment (Cypher1) [reply meta/peg.py:79]: Thanks, I added them due to running into errors but I think that was just a function of a parser fallback which is no longer happening due to other changes. Fixed

File: run_tests.py (hunk 2 above)
Comment (Cypher1) [reply run_tests.py]: Removed

**Review by orodley — APPROVED** [review]

**Comment by orodley** [comment]
Thanks for the contribution!

**Pull request merged**